README
NAME
    Plack::Middleware::TrafficLog - Log headers and body of HTTP traffic

SYNOPSIS
      # In app.psgi
      use Plack::Builder;

      builder {
          enable "TrafficLog", with_body => 1;
      };

DESCRIPTION
    This middleware logs the request and response messages with detailed
    information about headers and body.

    The example log:

      [08/Aug/2012:16:59:47 +0200] [164836368] [127.0.0.1 -> 0:5000] [Request ]
      |GET / HTTP/1.1|Connection: TE, close|Host: localhost:5000|TE: deflate,gzi
      p;q=0.3|User-Agent: lwp-request/6.03 libwww-perl/6.03||
      [08/Aug/2012:16:59:47 +0200] [164836368] [127.0.0.1 <- 0:5000] [Response]
      |HTTP/1.0 200 OK|Content-Type: text/plain||Hello World

    This module works also with applications which have delayed response. In
    that case each chunk is logged separately and shares the same unique ID
    number and headers.

    The body of request and response is not logged by default. For streaming
    responses only first chunk is logged by default.

SEE ALSO
    Plack, Plack::Middleware::AccessLog.

BUGS
    This module has unstable API and it can be changed in future.

    The log file can contain the binary data if the PSGI server provides
    binary files.

    If you find the bug or want to implement new features, please report it
    at http://github.com/dex4er/perl-Plack-Middleware-TrafficLog/issues

    The code repository is available at
    http://github.com/dex4er/perl-Plack-Middleware-TrafficLog

AUTHOR
    Piotr Roszatycki <dexter@cpan.org>

LICENSE
    Copyright (c) 2012, 2014 Piotr Roszatycki <dexter@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as perl itself.

    See http://dev.perl.org/licenses/artistic.html



Changes
0.0403  2023-06-18      Piotr Roszatycki <dexter@cpan.org>

* Correct configure requirements.

0.0402  2023-06-17      Piotr Roszatycki <dexter@cpan.org>

* Provides cpanfile.

0.0401  2015-10-14      Piotr Roszatycki <dexter@cpan.org>

* Bugfix for typo that shows some warnings on Perl 5.22.

0.0400  2014-06-28      Piotr Roszatycki <dexter@cpan.org>

* POSIX::strftime::Compiler replaces POSIX::strftime::GNU

0.0300  2014-06-28      Piotr Roszatycki <dexter@cpan.org>

* The body of request and response is not logged by default.
* New option: with_all_chunks. For streaming reponses only first chunk is
  logged by default.
* Requires POSIX::strftime::GNU

0.0202  2012-08-17      Piotr Roszatycki <dexter@cpan.org>

* Correct timezone on systems where strftime doesn't support %z.
* Log also REMOTE_PORT if is defined.

0.0201  2012-08-08      Piotr Roszatycki <dexter@cpan.org>

* Log responses without body.
* Log correctly responses with unknown status.

0.02    2012-08-07      Piotr Roszatycki <dexter@cpan.org>

* Delayed and streamed reponses are now supported.
* Build does not require Test::Exception.

0.01    2012-08-07      Piotr Roszatycki <dexter@cpan.org>

* First CPAN release.


Dist.SH
#!/usr/bin/env bash

set -e

PERL=${PERL:-perl}
PROVE=${PROVE:-prove}

test -x Build && ./Build distclean

${PERL} Build.PL

rm -f MANIFEST

${PERL} -ni -e 'print unless /^##? NAME/ .. eof' README.md
dist_version_from=$(${PERL} -ne 'print if /dist_version_from/' Build.PL | ${PERL} -pe "s%.*=> '%%; s%',%%")
${PERL} -MPod::Markdown -e '$p=Pod::Markdown->new;$p->parse_from_filehandle(\*STDIN);print $p->as_markdown' \
    <"${dist_version_from:-lib/$(${PERL} -ne 'next unless /^name: /; s/^name: //; s,-,/,g; s/$/.pm/; print' META.yml)}" |
    ${PERL} -pe 'sub a { $_[0]=~tr/ /-/; $_[0] } s,[(]http://search.cpan.org/perldoc[?]([^#)]*)(#[^)]*)?[)],a("(http://metacpan.org/module/$1$2)"),eg' |
    ${PERL} -pe 's/^# /## /' |
    ${PERL} -pe 's/^\s+```/```/' \
        >>README.md

${PERL} Build manifest

${PERL} Build distdir
diff -u -I^SIGNATURE MANIFEST ./*-[0-9]*.[0-9]*/MANIFEST

cd ./*-[0-9]*.[0-9]*

${PERL} Build.PL
${PERL} Build

${PROVE} -Iblib/lib t

if [[ -z ${TEST_SKIP_EXTRA} ]] || [[ ${TEST_SKIP_EXTRA} == 0 ]]; then
    ${PROVE} -Iblib/lib xt
elif [[ -n ${TEST_SKIP_EXTRA} ]] && [[ ${TEST_SKIP_EXTRA} != 1 ]]; then
    # shellcheck disable=SC2046
    ${PROVE} -Iblib/lib $(find xt -name '*.t' ! -name "${TEST_SKIP_EXTRA}")
fi

if [[ $1 == sign ]]; then
    ${PERL} Build distsign
    diff -u -I^SIGNATURE MANIFEST ./*-[0-9]*.[0-9]*/MANIFEST
else
    ${PERL} Build distdir
fi

chmod -R u+w,go-w ./*-[0-9]*.[0-9]*

export COPYFILE_DISABLE=true
tar zcf "$(basename "$(pwd)").tar.gz" ./*-[0-9]*.[0-9]*

mv ./*.tar.gz ..


Build.PL
#!/usr/bin/perl

use 5.006;

use strict;
use warnings;

use Module::Build;
use Module::CPANfile;

my $module = 'Plack::Middleware::TrafficLog';

(my $dist = $module) =~ s/::/-/g;

my $prereqs = Module::CPANfile->load->prereqs;

my $build = Module::Build->new(
    module_name             => $module,
    license                 => 'perl',
    build_requires          => $prereqs->requirements_for('build', 'requires')->as_string_hash,
    requires                => $prereqs->requirements_for('runtime', 'requires')->as_string_hash,
    recommends              => $prereqs->requirements_for('runtime', 'recommends')->as_string_hash,
    test_requires           => $prereqs->requirements_for('test', 'requires')->as_string_hash,
    auto_configure_requires => 0,
    create_makefile_pl      => 'traditional',
    create_readme           => 1,
    create_license          => 1,
    recursive_test_files    => 0,
    add_to_cleanup          => [
        '*.bak', '*.gz', '*.tgz', '*.ppd', '*.ppmx', '*.deb', '*.bat',
        'Makefile', 'pm_to_blib', 'cover_db', 'Debian_CPANTS.txt',
        "$dist-*",
    ],
    meta_merge => {
        resources => {
            bugtracker => "http://github.com/dex4er/perl-$dist/issues",
            repository => "http://github.com/dex4er/perl-$dist",
        },
    },
);

$build->create_build_script;


lib/Plack/Middleware/TrafficLog.pm
package Plack::Middleware::TrafficLog;

=head1 NAME

Plack::Middleware::TrafficLog - Log headers and body of HTTP traffic

=head1 SYNOPSIS

=for markdown ```perl

    # In app.psgi
    use Plack::Builder;

    builder {
        enable "TrafficLog", with_body => 1;
    };

=for markdown ```

=head1 DESCRIPTION

This middleware logs the request and response messages with detailed
information about headers and the body.

The example log:

=for markdown ```console

    [08/Aug/2012:16:59:47 +0200] [164836368] [127.0.0.1 -> 0:5000] [Request ]
    |GET / HTTP/1.1|Connection: TE, close|Host: localhost:5000|TE: deflate,gzi
    p;q=0.3|User-Agent: lwp-request/6.03 libwww-perl/6.03||
    [08/Aug/2012:16:59:47 +0200] [164836368] [127.0.0.1 <- 0:5000] [Response]
    |HTTP/1.0 200 OK|Content-Type: text/plain||Hello World

=for markdown ```

This module works also with applications that have delayed response. In that
case, each chunk is logged separately and shares the same unique ID number and
headers.

The body of the request and response is not logged by default. For streaming
responses, only the first chunk is logged by default.

=for readme stop

=cut

use 5.008;

use strict;
use warnings;

our $VERSION = '0.0403';

use parent 'Plack::Middleware';

use Plack::Util::Accessor qw(
    with_request with_response with_date with_body with_all_chunks eol body_eol logger
    _counter _call_id _strftime
);

use Plack::Util;

use Plack::Request;
use Plack::Response;

use POSIX                     ();
use POSIX::strftime::Compiler ();
use Scalar::Util              ();

sub prepare_app {
    my ($self) = @_;

    # the default values
    $self->with_request(Plack::Util::TRUE)                 unless defined $self->with_request;
    $self->with_response(Plack::Util::TRUE)                unless defined $self->with_response;
    $self->with_date(Plack::Util::TRUE)                    unless defined $self->with_date;
    $self->with_body(Plack::Util::FALSE)                   unless defined $self->with_body;
    $self->with_all_chunks(Plack::Util::FALSE)             unless defined $self->with_all_chunks;
    $self->body_eol(defined $self->eol ? $self->eol : ' ') unless defined $self->body_eol;
    $self->eol('|')                                        unless defined $self->eol;

    $self->_strftime(POSIX::strftime::Compiler->new('%d/%b/%Y:%H:%M:%S %z'));

    $self->_counter(0);
}

sub _log_message {    ## no critic(Subroutines::ProhibitManyArgs)
    my ($self, $type, $env, $status, $headers, $body) = @_;

    my $logger = $self->logger || sub { $env->{'psgi.errors'}->print(@_) };

    my $server_addr = sprintf '%s:%s', $env->{SERVER_NAME}, $env->{SERVER_PORT};
    my $remote_addr
        = defined $env->{REMOTE_PORT}
        ? sprintf '%s:%s', $env->{REMOTE_ADDR}, $env->{REMOTE_PORT}
        : $env->{REMOTE_ADDR};

    my $eol = $self->eol;
    my $body_eol = $self->body_eol;
    $body =~ s/\015?\012/$body_eol/gs if defined $body_eol;

    my $date
        = $self->with_date
        ? ('[' . $self->_strftime->to_string(localtime) . '] ')
        : '';

    $logger->(
        sprintf "%s[%s] [%s %s %s] [%s] %s%s%s%s%s%s\n",
        $date,
        $self->_call_id,

        $remote_addr,
        $type eq 'Request ' ? '->' : $type eq 'Response' ? '<-' : '--',
        $server_addr,

        $type,

        $eol,
        $status,
        $eol,
        $headers->as_string($eol),
        $eol,
        $body,
    );
}

sub _log_request {
    my ($self, $env) = @_;

    my $req = Plack::Request->new($env);

    my $status = sprintf '%s %s %s', $req->method, $req->request_uri, $req->protocol;
    my $headers = $req->headers;
    my $body = $self->with_body ? $req->content : '';

    $self->_log_message('Request ', $env, $status, $headers, $body);
}

sub _log_response {
    my ($self, $env, $ret) = @_;

    my $res = Plack::Response->new(@$ret);

    my $status_code = $res->status;
    my $status_message = HTTP::Status::status_message($status_code);

    my $status = sprintf 'HTTP/1.0 %s %s', $status_code, defined $status_message ? $status_message : '';
    my $headers = $res->headers;
    my $body = '';
    if ($self->with_body) {
        $body = $res->content;
        $body = '' unless defined $body;
        $body = join '', grep { defined $_ } @$body if ref $body eq 'ARRAY';
    }

    $self->_log_message('Response', $env, $status, $headers, $body);
}

sub call {
    my ($self, $env) = @_;

    $self->_call_id(
        sprintf '%015d',
        time % 2**16
            * 2**32
            + (Scalar::Util::looks_like_number $env->{REMOTE_PORT} ? $env->{REMOTE_PORT} : int rand 2**16) % 2**16
            * 2**16
            + $self->_counter % 2**16
    );
    $self->_counter($self->_counter + 1);

    # Preprocessing
    $self->_log_request($env) if $self->with_request;

    # $self->app is the original app
    my $res = $self->app->($env);

    # Postprocessing
    return $self->with_response
        ? $self->response_cb(
        $res,
        sub {
            my ($ret) = @_;
            my $seen;
            return sub {
                my ($chunk) = @_;
                return        if $seen and not defined $chunk;
                return $chunk if $seen and not $self->with_all_chunks;
                $self->_log_response($env, [$ret->[0], $ret->[1], [$chunk]]);
                $seen = Plack::Util::TRUE;
                return $chunk;
            };
        }
        )
        : $res;
}

1;

__END__

=head1 CONFIGURATION

=over 4

=item logger

Sets a callback to print log message to. It prints to C<psgi.errors> output
stream by default.

=for markdown ```ini

    # traffic.l4p
    log4perl.logger.traffic = DEBUG, LogfileTraffic
    log4perl.appender.LogfileTraffic = Log::Log4perl::Appender::File
    log4perl.appender.LogfileTraffic.filename = traffic.log
    log4perl.appender.LogfileTraffic.layout = PatternLayout
    log4perl.appender.LogfileTraffic.layout.ConversionPattern = %m{chomp}%n

=for markdown ```

=for markdown ```perl

    # app.psgi
    use Log::Log4perl qw(:levels get_logger);
    Log::Log4perl->init('traffic.l4p');
    my $logger = get_logger('traffic');

    enable "Plack::Middleware::TrafficLog",
        logger => sub { $logger->log($INFO, join '', @_) };

=for markdown ```

=item with_request

The false value disables logging of the request message.

=item with_response

The false value disables logging of the response message.

=item with_date

The false value disables logging of the current date.

=item with_body

The true value enables logging of the message's body.

=item with_all_chunks

The true value enables logging of every chunk for streaming responses.

=item eol

Sets the line separator for the message's headers and the body. The default
value is the pipe character C<|>.

=item body_eol

Sets the line separator for message's body only. The default is the space
character C< >. The default value is used only if B<eol> is also undefined.

=back

=for readme continue

=head1 SEE ALSO

L<Plack>, L<Plack::Middleware::AccessLog>.

=head1 BUGS

This module has an unstable API and it can be changed in the future.

The log file can contain binary data if the PSGI server provides binary
files.

If you find the bug or want to implement new features, please report it at
L<http://github.com/dex4er/perl-Plack-Middleware-TrafficLog/issues>

The code repository is available at
L<http://github.com/dex4er/perl-Plack-Middleware-TrafficLog>

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Copyright (c) 2012, 2014-2015, 2023 Piotr Roszatycki <dexter@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as perl itself.

See L<http://dev.perl.org/licenses/artistic.html>
